driver/mloCartesianProdHost.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <miopen/tensor.hpp>
+#include <miopen/tensor_view_utils.hpp>
+#include <../test/ford.hpp>
+
+template <typename Tgpu, typename Tcheck>
+int32_t mloCartesianProdForwardRunHost(std::vector<miopenTensorDescriptor_t> inputDescs,
+                                       const miopenTensorDescriptor_t outputDesc,
+                                       const std::vector<Tgpu*> inputs,
+                                       Tcheck* output)
+{
+    auto output_tv = miopen::get_inner_expanded_tv<2>(miopen::deref(outputDesc));
+
+    Tcheck* output_ws = new Tcheck[output_tv.size[0] * output_tv.size[1]];
+    size_t stride     = 1;
+    for(int dim1_idx = inputs.size() - 1; dim1_idx >= 0; --dim1_idx)
+    {
+        auto input_tv = miopen::get_inner_expanded_tv<1>(miopen::deref(inputDescs[dim1_idx]));
+        auto numel    = miopen::deref(outputDesc).GetLengths()[0];
+
+        ford(numel)([&](size_t gid) {
+            if(gid >= output_tv.size[0])
+                return;
+
+            output_ws[output_tv.size[0] * dim1_idx + gid] =
+                inputs[dim1_idx][(gid / stride) % input_tv.size[0]];
+        });
+        stride *= miopen::deref(inputDescs[dim1_idx]).GetElementSize();
+    }
+
+    par_ford(output_tv.size[1])([&](size_t dim1_idx) {
+        ford(output_tv.size[0])([&](size_t dim0_idx) {
+            output[output_tv.get_tensor_view_idx({dim0_idx, dim1_idx})] =
+                output_ws[output_tv.size[0] * dim1_idx + dim0_idx];
+        });
+    });
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tcheck>
+int32_t mloCartesianProdBackwardRunHost(const miopenTensorDescriptor_t outputGradDesc,
+                                        std::vector<miopenTensorDescriptor_t> inputGradDescs,
+                                        const Tgpu* output_grad,
+                                        std::vector<Tcheck*>& input_grads)
+{
+    auto output_grad_tv = miopen::get_inner_expanded_tv<2>(miopen::deref(outputGradDesc));
+    size_t stride       = 1;
+    for(int dim1_idx = input_grads.size() - 1; dim1_idx >= 0; --dim1_idx)
+    {
+        auto input_grad_tv =
+            miopen::get_inner_expanded_tv<1>(miopen::deref(inputGradDescs[dim1_idx]));
+        auto numel = miopen::deref(inputGradDescs[dim1_idx]).GetElementSize();
+
+        ford(numel)([&](size_t gid) {
+            if(gid >= input_grad_tv.size[0])
+                return;
+            float sum = 0;
+            for(size_t offset = 0; offset < output_grad_tv.size[0];
+                offset += (stride * input_grad_tv.size[0]))
+            {
+                for(size_t i = 0; i < stride; ++i)
+                {
+                    size_t dim0_idx = offset + gid * stride + i;
+                    sum += static_cast<float>(
+                        output_grad[output_grad_tv.get_tensor_view_idx({dim0_idx, dim1_idx})]);
+                }
+            }
```

Comment:
Use `par_ford` for better performance
```suggestion
            std::vector<double> buffer(output_grad_tv.size[0], 0);
            // apply par_ford
            for(size_t offset = 0; offset < output_grad_tv.size[0];
                offset += (stride * input_grad_tv.size[0]))
            {
                for(size_t i = 0; i < stride; ++i)
                {
                    size_t dim0_idx = offset + gid * stride + i;
                    buffer[offset] += static_cast<float>(
                        output_grad[output_grad_tv.get_tensor_view_idx({dim0_idx, dim1_idx})]);
                }
            }
            auto sum = std::accumulate(buffer.begin(), buffer.end(), 0.0);
```

Reply:
resolved